src/main/java/com/ontotext/kafka/service/AddRecordsProcessor.java

```diff
@@ -19,26 +27,69 @@
  */
 public class AddRecordsProcessor extends SinkRecordsProcessor {
 
-	public AddRecordsProcessor(Queue<Collection<SinkRecord>> sinkRecords, AtomicBoolean shouldRun,
-			Repository repository, RDFFormat format, int batchSize, long timeoutCommitMs) {
-		super(sinkRecords, shouldRun, repository, format, batchSize, timeoutCommitMs);
+	public AddRecordsProcessor(Queue<Collection<SinkRecord>> sinkRecords, AtomicBoolean shouldRun, Repository repository,
+							   RDFFormat format, int batchSize, long timeoutCommitMs, ErrorHandler errorHandler) {
+		super(sinkRecords, shouldRun, repository, format, batchSize, timeoutCommitMs, errorHandler);
 	}
 
 	@Override
 	public void flushRecordUpdates() {
-		//no need to create connection if already empty
-		if (!recordsBatch.isEmpty()) {
+		try {
+			if (!recordsBatch.isEmpty()) {
+				flush();
+				recordsBatch.clear();
+			}
+		} catch (TimeoutException e) {
+			LogErrorHandler.logError(e.getMessage());
+			throw new RuntimeException(e);
+		}
+	}
+
+	private void flush() throws TimeoutException {
+		long retryCount = 1L;
+		Queue<SinkRecord> failedRecords = new ArrayDeque<>(recordsBatch.size());
+
+		do {
+			if (!failedRecords.isEmpty()) {
+				removeFailed(failedRecords);
+			}
+
+			long startAttempt = System.currentTimeMillis();
 			try (RepositoryConnection connection = repository.getConnection()) {
 				connection.begin();
-				while (recordsBatch.peek() != null) {
-					connection.add(ValueUtil.convertRDFData(recordsBatch.poll().value()), format);
+				for (SinkRecord record : recordsBatch) {
+					addRecord(record, connection, failedRecords);
 				}
 				connection.commit();
-			} catch (IOException e) {
-				throw new RuntimeException(e);
-				//todo first add retries
-				//todo inject error handler
+				return;
+
+			} catch (IOException | RepositoryException e) {
+				LogErrorHandler.logWarning("Exception while flushing records " + e.getMessage());
+				//retry sending all valid records in the batch until successful or timeout
+				while (DEFERRED_TIME_BETWEEN_RETRIES > System.currentTimeMillis() - startAttempt) {
+					//do nothing
+				}
+				if (NUMBER_OF_CONNECTION_RETRIES <= retryCount) {
+					throw new TimeoutException("Flushing run out of time due to: " + e.getMessage());
+				}
+				LogErrorHandler.logInfo("Retrying connection...");
+				retryCount++;
 			}
+		} while (true);
```

Comment:
never a good idea to have a while (true) block, why don't you use the retries count here